brainiak/matnorm_models/regression.py

```diff
@@ -0,0 +1,167 @@
+import tensorflow as tf
+import numpy as np
+from .helpers import define_scope
+from .base import MatnormModelBase
+
+
+class MatnormRegression(MatnormModelBase):
+    """ This analysis allows maximum likelihood estimation of regression models
+    in the presence of both spatial and temporal covariance.
+
+    ..math::
+    Y \\sim \\mathcal{MN}(X\beta, time_noise_cov, space_noise_cov)
+
+    Parameters
+    ----------
+    n_c : int
+        Number of columns in design matrix
+    n_v : int
+        number of voxels
+    time_noise_cov : subclass of NoiseCovBase
+        TR noise covariance class following NoiseCovBase interface.
+    space_noise_cov : subclass of NoiseCovBase
+        Voxel noise covariance class following NoiseCovBase interface.
+    learnRate : real, default=0.01
+        Step size for the Adam optimizer
+
+    """
+    def __init__(self, n_c, n_v, time_noise_cov, space_noise_cov,
+                 learnRate=0.01):
+
+        self.n_c, self.n_v, self.time_noise_cov, self.space_noise_cov =\
```

Comment:
I've intentionally avoided having a base class `__init__` to be maximally agnostic about derived models -- right now the base class is basically a collection of methods without any state. For example, if we were to implement SRM in this framework we wouldn't have time and space cov, we'd have subject-space and time covs, and we wouldn't have n_c. If we were to replace the optimizer with the "flip-flop" coordinate ascent algo for matrix normal covariance estimation, we wouldn't have the same optimizer node -- and so on and so forth. I do want to make more commitments to the base API, but want to do them thoughtfully based on more than just these two simple models. Maybe I'm overthinking it at this early stage, though?